### ISSUE TITLE
index full of duplicates (no unique constraint?)

### ISSUE THREAD

**Issue by bradfordcondon** [statonlab/tripal_cv_xray#39]
select * from the index where db/accession = something

for example:

```
4227376	73928	GO	0051179	132373
4227377	73928	GO	0051179	132373
4227378	73928	GO	0051179	132373
4227379	73928	GO	0051179	132373
4227380	73928	GO	0051179	132373
4227381	73928	GO	0051179	132373
4227382	73928	GO	0051179	132373
4227383	73928	GO	0051179	132373
4227384	73928	GO	0051179	132373
4227385	73928	GO	0051179	132373
4227386	73928	GO	0051179	132373
4227387	73928	GO	0051179	132373
```

this is from the cvterm_entity_linker table on live.  Here's [the corresponding entity](http://hardwoodgenomics.org/bio_data/132373)

**Comment by bradfordcondon** [comment]
I count 9 GO annotations, resulting in 12 links to term GO: 0051179 (which is not directly linked).

bizarrely enough go: 0051179 doesnt even appear to be a term.

![screen shot 2018-06-19 at 10 18 50 am](https://user-images.githubusercontent.com/7063154/41603055-2f8f04a8-73aa-11e8-9371-a440dd1a1cf6.png)

**Comment by almasaeed2010** [comment]
So the fix I added in #42 makes sure the indexer doesn't insert the same record more than once. It seems to work on my local dev and I am not really sure how much slower it will be but it shouldn't be that bad given that we have a queue system in place. Also note that the PR does not add the constraints on the db table which might not be absolutely necessary since only the indexer inserts data to the table and it already does the checks.

**Comment by bradfordcondon** [comment]
ok great.  agreed it will be hard to test: Ill look at code, merge, and rerun on live ince thats where i see the problem and we wont be breaking any functionality.

**Comment by bradfordcondon** [comment]
not resolved with #42 unfortunately.

<img width="636" alt="screen shot 2018-06-26 at 3 54 03 pm" src="https://user-images.githubusercontent.com/7063154/41935867-4731ad40-7959-11e8-84d6-8d6d61c098ea.png">

**Comment by bradfordcondon** [comment]
i created a test case on the dupe branch that you can run .  takes me ~1 minute to run.  You need to load the GOSLIM OBO first.

```
  /**
   * Checks of a row already exists.
   *
   * @param $row
   *
   * @return bool
   */
  public function exists($row) {
    $query = db_select('tripal_cvterm_entity_linker', 'TCEL');

    foreach ($row as $field => $value) {
      var_dump($field);
      var_dump($value);

      $query->condition($field, $value);
    }

    var_dump($query->countQuery()->execute()->fetchField());

    return ((int) $query->countQuery()->execute()->fetchField()) > 0;
  }
```


so for some reason this query is always 0. Dumping each condition gives me strings of stuff like this:


```

string(9) "entity_id"
string(5) "34065"
string(9) "cvterm_id"
string(4) "2939"
string(8) "database"
string(2) "GO"
string(9) "accession"
string(7) "0003676"
string(1) "0"
string(9) "entity_id"
string(5) "34065"
string(9) "cvterm_id"
string(4) "2939"
string(8) "database"
string(2) "GO"
string(9) "accession"
string(7) "0003676"
string(1) "0"
string(9) "entity_id"
string(5) "34065"
string(9) "cvterm_id"
string(4) "2943"
string(8) "database"
string(2) "GO"
string(9) "accession"
string(7) "0003723"
string(1) "0"
````

You can see that evne in this string of var_dumps you have the same entity/cvterm being associated.

It seems clear to me why: `insertData` only inserts ONCE, at the end.  So, this `exists` check will *always* be false, i think.  We need to either 
*  insert at each step, or
* comb through the built query at the end and remove duplicate combinations.

**Comment by almasaeed2010** [comment]
I pushed a new set of changes that guarantees no duplication happens. The new changes have passed all your unit tests. I also did manual testing of entity insert hook by creating a new entity. It would be nice to have a unit test to test the hooks though and since the publish method in TripalTestSuite does trigger the hook, the same method used to test dupes should work. We'd only need to remove the part where you manually trigger the handle method of the indexer.

**Comment by bradfordcondon** [comment]
agreed.  it actually doesnt matter which method we confirm doesnt have duplicates (or, we could test both).

EIther way, the work is almost done to have test coverage of both the hooks **and** the job so I'm pretty pleased/ think it was worth the time to make the case.  Let's push it over the finish line and explicitly test each case, hook and index job.

**Comment by almasaeed2010** [comment]
Implemented a unit test for hook entity insert (see https://github.com/statonlab/tripal_cv_xray/blob/hooks_and_prev_indexing/tests/XrayIndexerJobTest.php).

**Comment by almasaeed2010** [comment]
Implemented a unit test for the update hook (same file as above). Everything is pushed and ready to merged. 

@bradfordcondon I'll create the PR for your review.

**Comment by bradfordcondon** [comment]
should be resolved with #47

**Comment by bradfordcondon** [comment]
looks reasonable.  ill update it and index live as well.
